capture/capture.proto

```diff
@@ -129,63 +234,147 @@ message ValidateResponse {
+  bool rfc7396_merge_patch = 2;
 }
 
-// CaptureResponse is the response type of a Capture RPC.
+// Acknowledge is a notification that a Checkpoint has committed to the
+// Flow runtime's recovery log.
+message Acknowledge {}
+
+// PullRequest is the request type of a Driver.Pull RPC.
 // It will have exactly one top-level field set, which represents its message
 // type.
-message CaptureResponse {
-  // Opened responds to CaptureRequest of the client,
-  // and is sent exactly once as the first message of the stream.
-  message Opened {
+message PullRequest {
+  // Open opens a Pull of the driver, and is sent exactly once as the first
+  // message of the stream.
+  message Open {
+    // CaptureSpec to be pulled.
+    flow.CaptureSpec capture = 1;
+    // [key_begin, key_end] inclusive range of keys processed by this
+    // transaction stream. Ranges reflect the disjoint chunks of ownership
+    // specific to each instance of a scale-out capture implementation.
+    fixed32 key_begin = 2;
+    fixed32 key_end = 3;
+    // Last-persisted driver checkpoint from a previous capture stream.
+    // Or empty, if the driver has cleared or never set its checkpoint.
+    bytes driver_checkpoint_json = 4 [
+      (gogoproto.casttype) = "encoding/json.RawMessage",
+      json_name = "driverCheckpoint"
+    ];
+    // If true, perform a blocking tail of the capture.
+    // If false, produce all ready output and then close the stream.
+    bool tail = 5;
   }
+  Open open = 1;
+
+  // Tell the driver that its Checkpoint has committed to the Flow recovery log.
+  Acknowledge acknowledge = 2;
+}
+
+// PullResponse is the response type of a Driver.Pull RPC.
+// It will have exactly one top-level field set, which represents its message
+// type.
+message PullResponse {
+  // Opened responds to PullRequest.Open of the runtime,
+  // and is sent exactly once as the first message of the stream.
+  message Opened {}
   Opened opened = 1;
 
-  // Captured returns documents of the capture stream.
-  message Captured {
-    // The capture binding for documents of this Captured response.
-    uint32 binding = 1;
-    // Byte arena of the response.
-    bytes arena = 2 [
-      (gogoproto.casttype) = "github.com/estuary/protocols/flow.Arena"
-    ];
-    // Captured JSON documents.
-    repeated flow.Slice docs_json = 3 [ (gogoproto.nullable) = false ];
+  // Captured documents of the stream.
+  Documents documents = 2;
+
+  // Checkpoint all preceeding Documents of this stream.
+  Checkpoint checkpoint = 3;
+}
+
+// PushRequest is the request message of the Runtime.Push RPC.
+message PushRequest {
+  // Open opens a Push of the runtime, and is sent exactly once as the first
+  // message of the stream.
+  message Open {
+    // Header identifies a specific Shard and Route to which this stream is
+    // directed. It's optional, and is typically attached by a proxying peer.
+    protocol.Header header = 1;
```

Comment:
protocol.Header is a Gazette-ism, used to dynamically route and verify redirected requests to specific members. It's tightly integrated into consumer.Resolver: [ResolverArgs](https://pkg.go.dev/go.gazette.dev/core@v0.89.0/consumer#ResolveArgs) takes a proxied header, and [Resolution](https://pkg.go.dev/go.gazette.dev/core@v0.89.0/consumer#Resolution) will give you one that can be attached to requests. Headers are essentially a by-product of doing shard (or journal) resolution, and can be attached to request to essentially declare a routing outcome that you believe should be true, and which the peer can verify.

The broker's [resolver.go](https://github.com/gazette/core/blob/master/broker/resolver.go) works identically, but isn't exported because it's an internal implementation detail of the broker. `consumer` exports it to allow for dynamic routing in application-specific RPCs like this one.

As a practical matter, this isn't a field that the caller must or even should set. It's attached by an instance of the receiving service which is proxying the request to a peer on the caller's behalf.

Here's the definition of a [Header](https://github.com/gazette/core/blob/master/broker/protocol/protocol.proto#L680-L714). Briefly, it's the state of Etcd (revision to read through); the identified ProcessID, and the routing topology of the item (shard ID, in this context) in question.